### PR TITLE
Implement R8 shrinker

### PR DESCRIPTION
**Pull request by PranavPurwar**

### CODE REVIEW

File: app/src/minApi21/java/mod/hey/studios/build/BuildSettingsDialogBridge.java [hunk 1]
```diff
@@ -11,8 +12,13 @@
 public class BuildSettingsDialogBridge {
     public static String[] getAvailableJavaVersions() {
         return new String[] {
-                SETTING_JAVA_VERSION_1_7,
-                SETTING_JAVA_VERSION_1_8,
+                SETTING_JAVA_VERSION_1_7
```
Comment (JavkhlanK) [inline app/src/minApi21/java/mod/hey/studios/build/BuildSettingsDialogBridge.java:15]: Please at least make a separate commit for this change.

File: app/src/main/java/mod/pranav/build/JarBuilder.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                if (count == -1) break;
+                target.write(buffer, 0, count);
+            }
+            target.closeEntry();
+
+        } finally {
+            if (in != null) {
+                in.close();
+            }
+        }
```
Comment (JavkhlanK) [inline app/src/main/java/mod/pranav/build/JarBuilder.java:81]: Can't you use a try-with-resource instead?

**Review by JavkhlanK — CHANGES_REQUESTED** [review]
Looks all-in-all neat, but to me, having R8/ProGuard in Build Settings is a bad UI decision. I think it should be adjustable in ProGuard Manager instead.

**Comment by PranavPurwar** [comment]
> Looks all-in-all neat, but to me, having R8/ProGuard in Build Settings is a bad UI decision. I think it should be adjustable in ProGuard Manager instead.

I did this in a hurry, didn't really think of that.

File: app/src/minApi21/java/mod/hey/studios/build/BuildSettingsDialogBridge.java (hunk 1 above)
Comment (PranavPurwar) [reply app/src/minApi21/java/mod/hey/studios/build/BuildSettingsDialogBridge.java:15]: you couldn't compile java 8 in minApi21 anyways

**Comment by JavkhlanK** [comment]
bro sir what happened in the last commit

**Comment by ilyassesalama** [comment]
> Looks all-in-all neat, but to me, having R8/ProGuard in Build Settings is a bad UI decision. I think it should be adjustable in ProGuard Manager instead.

To be honest, the whole build dialog is a bad UI decision. IMO, when the user clicks on "build settings" button, an activity should appear showing all possible toggles or modifications that can be made, and even shortcuts to take you to specific activities such as ProGuard settings.

You may ask, why I have chosen an activity? It is because the current UI has become so crowded, and future features will not fit there well. So, it's better to take a step ahead to solve this issue permanently.
You may also ask, how would the activity look? In my opinion, I think it should look like AndroidManifest Manager screen.

**Comment by PranavPurwar** [comment]
no use fragments instead for pagination in different build options like `Compiler Options`, `Shrinker options`, etc.

**Comment by DereckySany** [comment]
> > Looks all-in-all neat, but to me, having R8/ProGuard in Build Settings is a bad UI decision. I think it should be adjustable in ProGuard Manager instead.
> 
> To be honest, the whole build dialog is a bad UI decision. IMO, when the user clicks on "build settings" button, an activity should appear showing all possible toggles or modifications that can be made, and even shortcuts to take you to specific activities such as ProGuard settings.
> 
> You may ask, why I have chosen an activity? It is because the current UI has become so crowded, and future features will not fit there well. So, it's better to take a step ahead to solve this issue permanently. You may also ask, how would the activity look? In my opinion, I think it should look like AndroidManifest Manager screen.

in my opinion, "build settings" should stay the way it is offering unique and quick options to be adjusted, and not creating an activity full of options that already exist in their own sections,for advanced sections we have the drawer on the right, last position where you can add anything for advanced controls,without messing up the sketchware interface that everyone is used to.

**Comment by PranavPurwar** [comment]
merge now

**Comment by moderGamer** [comment]
Hi, does this feature encrypt the project like ProGuard?

**Comment by PranavPurwar** [comment]
> Hi, does this feature encrypt the project like ProGuard?

@moderGamer yes, but its much more efficient and better

**Comment by moderGamer** [comment]
> 

Very good, so fix the error so that it is added to the new sketchware feature, thank you

**Comment by PranavPurwar** [comment]
> > 
> 
> Very good, so fix the error so that it is added to the new sketchware feature, thank you

@moderGamer its the upstream changes causing issue

**Comment by moderGamer** [comment]
> > > 
> > 
> > 
> > Very good, so fix the error so that it is added to the new sketchware feature, thank you
> 
> @moderGamer its the upstream changes causing issue

Good luck building this feature

**Comment by moderGamer** [comment]
Screenshot_20230221-232225_Dexter

@PranavPurwar Hello are you feeling good?  I tested the last action and created a program, as you can see in the photo, it did not block a specific code. Can you explain a little about this feature, thanks?

**Comment by PranavPurwar** [comment]
@moderGamer did you enable proguard in settings?

**Comment by moderGamer** [comment]
> 

no bro

**Comment by PranavPurwar** [comment]
> > 
> 
> no bro

you need to enable that to start the shrinker

**Comment by PranavPurwar** [comment]
> 

Are you on less than API 26 device?

**Comment by moderGamer** [comment]
> > 
> 
> Are you on less than API 26 device?

Screenshot_20230222-092644_Sketchware Pro


is active

**Comment by TheRemakerMan** [comment]
> > 
> 
> Are you on less than API 26 device?

No, I have a device with the 30 version.

**Comment by moderGamer** [comment]
> 

my android version 11

**Comment by TheRemakerMan** [comment]
> > 
> 
> my android version 11
> 
> @TheRemakerMan

I didn't talk to you, bruh.

**Comment by PranavPurwar** [comment]
@TheRemakerMan you downloaded the minApi21 variant, its only available on minApi26 variant due to NIO dependency of R8/D8

**Comment by TheRemakerMan** [comment]
> @TheRemakerMan you downloaded the minApi21 variant, its only available on minApi26 variant due to NIO dependency of R8/D8

Oh, thanks.

**Comment by moderGamer** [comment]
> @TheRemakerMan you downloaded the minApi21 variant, its only available on minApi26 variant due to NIO dependency of R8/D8

I also installed the same version, it does not lock a special code

**Comment by PranavPurwar** [comment]
@moderGamer Enable proguard from the drawer and enable R8 from build settings dialog and build the project

**Comment by moderGamer** [comment]
> 

Aha ok bro tnks to Help me
